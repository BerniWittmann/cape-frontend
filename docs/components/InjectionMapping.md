# InjectionMapping

A Component to edit an Injection Mapping

## Props

<!-- @vuese:InjectionMapping:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|injectionMapping|the Injection Mapping|`Object`|`true`|-|

<!-- @vuese:InjectionMapping:props:end -->


