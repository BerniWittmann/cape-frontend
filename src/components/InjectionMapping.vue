<template>
  <p>TODO</p>
</template>

<script>

/*
 * @vuese
 * @group Components
 *
 * A Component to edit an Injection Mapping
 */
export default {
  name: 'InjectionMapping',
  props: {
    // the Injection Mapping
    injectionMapping: {
      type: Object,
      required: true
    }
  }
}

</script>

<style lang="scss">

</style>
