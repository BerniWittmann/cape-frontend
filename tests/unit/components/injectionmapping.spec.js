import { mount } from '@vue/test-utils'
import { i18n, monkeyPatchTransitions } from '../setupPlugins'

monkeyPatchTransitions()

import InjectionMapping from '@/components/InjectionMapping.vue'
import InjectionMappingService from '@/services/injectionMapping'
import InjectionMappingModel from '@/models/injectionMapping'

describe('Components', () => {
  describe('InjectionMapping', () => {
    let propsData
    let cmp
    let store
    InjectionMappingService.update = jest.fn()
    console.warn = jest.fn()

    beforeEach(() => {
      propsData = {
        injectionMapping: {
          id: '142',
          processID: '42',
          extensionAreaID: 'EA_1',
          injectedProcess: undefined,
          contextSituation: {
            id: 'c1',
            name: 'My Context Situation 1'
          }
        }
      }
      store = {
        getters: {
          'contextSituation/contextSituationsByTags': [{
            value: '0',
            label: 'untagged',
            children: []
          }, {
            value: 't1',
            label: 'Tag 1',
            children: [{
              key: 'c1',
              label: 'My Context Situation 1',
              value: {
                id: 'c1',
                name: 'My Context Situation 1'
              }
            }, {
              key: 'c2',
              label: 'My Context Situation 2',
              value: {
                id: 'c2',
                name: 'My Context Situation 2'
              }
            }]
          }]
        }
      }
      render()
    })

    function render() {
      cmp = mount(InjectionMapping, {
        i18n,
        propsData,
        mocks: {
          $store: store
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders a cascader for the context situation', () => {
      expect(cmp.contains('.el-cascader')).toBeTruthy()
    })

    it('preselects the context situation in the cascader', () => {
      expect(cmp.vm.currentContextSituation).toEqual(['t1', { 'id': 'c1', 'name': 'My Context Situation 1' }])
      const input = cmp.find('.el-cascader__label')
      expect(input.text()).toEqual('My Context Situation 1')
    })

    it('can change the context situation', () => {
      const cascader = cmp.find('.el-cascader')
      cascader.trigger('click')

      expect(cmp.html()).toMatchSnapshot()
      cmp.vm.currentContextSituation = ['0', {
        id: 'c2',
        name: 'My Context Situation 2'
      }]
      expect(cmp.html()).toMatchSnapshot()
      const input = cmp.find('.el-cascader__label')
      expect(input.text()).toEqual('My Context Situation 2')
    })

    it('has a save button', () => {
      expect(cmp.contains('.el-button--success')).toBeTruthy()
    })

    it('can save the injection specification', () => {
      const button = cmp.find('.el-button--success')
      button.trigger('click')

      expect(InjectionMappingService.update).toHaveBeenCalledWith(new InjectionMappingModel({
        '_id': '142',
        'context_situation': { '_id': 'c1', 'name': 'My Context Situation 1', 'rules': undefined, 'tags': [] },
        'extension_area_id': 'EA_1',
        'injected_process': {
          '_id': undefined,
          'description': undefined,
          'name': undefined,
          'svg': undefined,
          'tags': [],
          'xml': undefined
        },
        'process_id': '42'
      }))
    })

    describe('it has no context Situation', () => {
      beforeEach(() => {
        propsData.injectionMapping.contextSituation = undefined
        jest.clearAllMocks()
        render()
      })

      it('does not select the context Situation', () => {
        expect(cmp.vm.currentContextSituation).toEqual([])
        const input = cmp.find('.el-cascader__label')
        expect(input.text()).toEqual('')
      })

      it('shows a required error on save', () => {
        const button = cmp.find('.el-button--success')
        button.trigger('click')

        expect(InjectionMappingService.update).not.toHaveBeenCalled()
        expect(cmp.html()).toMatchSnapshot()
        const err = cmp.find('.el-form-item__error')
        expect(err.exists()).toBeTruthy()
        expect(err.text()).toEqual('injection_mapping.validation.context_situation.required')
      })

      it('does not find the context Situation', () => {
        propsData.injectionMapping.contextSituation = {
          id: 'c999',
          name: 'My not found CS'
        }
        render()

        expect(cmp.vm.currentContextSituation).toEqual([])
        const input = cmp.find('.el-cascader__label')
        expect(input.text()).toEqual('')
      })
    })
  })
})
