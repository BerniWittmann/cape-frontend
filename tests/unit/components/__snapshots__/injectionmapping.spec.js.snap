// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components InjectionMapping it can handle a context situation can change the context situation 1`] = `
<div class="el-form-item is-required"><label for="contextSituation" class="el-form-item__label">injection_mapping.context_situation</label>
  <div class="el-form-item__content"><span class="el-cascader is-opened"><div class="el-input el-input--suffix is-focus"><!----><input type="text" autocomplete="off" class="el-input__inner" aria-expanded="true"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down is-reverse"></i><!----></span>
    <!----></span>
    <!---->
  </div><span class="el-cascader__label">
      My Context Situation 1
    </span></span>
  <!---->
</div>
</div>
`;

exports[`Components InjectionMapping it can handle a context situation can change the context situation 2`] = `
<div class="el-form-item is-required"><label for="contextSituation" class="el-form-item__label">injection_mapping.context_situation</label>
  <div class="el-form-item__content"><span class="el-cascader is-opened"><div class="el-input el-input--suffix is-focus"><!----><input type="text" autocomplete="off" class="el-input__inner" aria-expanded="true" value="My Context Situation 2"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down is-reverse"></i><!----></span>
    <!----></span>
    <!---->
  </div><span class="el-cascader__label">
      My Context Situation 2
    </span></span>
  <!---->
</div>
</div>
`;

exports[`Components InjectionMapping it can handle a context situation it has no context Situation shows a required error on save 1`] = `
<div class="el-form-item is-error is-required"><label for="contextSituation" class="el-form-item__label">injection_mapping.context_situation</label>
  <div class="el-form-item__content"><span class="el-cascader"><div class="el-input el-input--suffix"><!----><input type="text" readonly="readonly" autocomplete="off" placeholder="injection_mapping.context_situation" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down"></i><!----></span><i class="el-input__icon el-input__validateIcon el-icon-loading"></i></span>
    <!---->
  </div><span class="el-cascader__label">

    </span></span>
  <div class="el-form-item__error" name="el-zoom-in-top">
    injection_mapping.validation.context_situation.required
  </div>
</div>
</div>
`;

exports[`Components InjectionMapping it can handle a injected process can change the injected process 1`] = `
<div class="el-form-item is-required"><label for="injectedProcess" class="el-form-item__label">injection_mapping.injected_process</label>
  <div class="el-form-item__content"><span class="el-cascader is-opened"><div class="el-input el-input--suffix is-focus"><!----><input type="text" autocomplete="off" class="el-input__inner" aria-expanded="true"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down is-reverse"></i><!----></span>
    <!----></span>
    <!---->
  </div><span class="el-cascader__label">
      Injected Process 1
    </span></span>
  <!---->
</div>
</div>
`;

exports[`Components InjectionMapping it can handle a injected process can change the injected process 2`] = `
<div class="el-form-item is-required"><label for="injectedProcess" class="el-form-item__label">injection_mapping.injected_process</label>
  <div class="el-form-item__content"><span class="el-cascader is-opened"><div class="el-input el-input--suffix is-focus"><!----><input type="text" autocomplete="off" class="el-input__inner" aria-expanded="true" value="Injected Process 2"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down is-reverse"></i><!----></span>
    <!----></span>
    <!---->
  </div><span class="el-cascader__label">
      Injected Process 2
    </span></span>
  <!---->
</div>
</div>
`;

exports[`Components InjectionMapping it can handle a injected process it has no injected process shows a required error on save 1`] = `
<div class="el-form-item is-error is-required"><label for="injectedProcess" class="el-form-item__label">injection_mapping.injected_process</label>
  <div class="el-form-item__content"><span class="el-cascader"><div class="el-input el-input--suffix"><!----><input type="text" readonly="readonly" autocomplete="off" placeholder="injection_mapping.injected_process" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down"></i><!----></span><i class="el-input__icon el-input__validateIcon el-icon-loading"></i></span>
    <!---->
  </div><span class="el-cascader__label">

    </span></span>
  <div class="el-form-item__error" name="el-zoom-in-top">
    injection_mapping.validation.injected_process.required
  </div>
</div>
</div>
`;

exports[`Components InjectionMapping renders 1`] = `
<div>
  <form class="el-form injection-mapping-form el-form--label-top el-form--inline">
    <div class="el-form-item">
      <!---->
      <div class="el-form-item__content">
        <p class="ifft-text">injection_mapping.if</p>
        <!---->
      </div>
    </div>
    <div class="el-form-item is-required"><label for="contextSituation" class="el-form-item__label">injection_mapping.context_situation</label>
      <div class="el-form-item__content"><span class="el-cascader"><div class="el-input el-input--suffix"><!----><input type="text" readonly="readonly" autocomplete="off" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down"></i><!----></span>
        <!----></span>
        <!---->
      </div><span class="el-cascader__label">
      My Context Situation 1
    </span></span>
      <!---->
    </div>
</div>
<div class="el-form-item">
  <!---->
  <div class="el-form-item__content">
    <p class="ifft-text">injection_mapping.then</p>
    <!---->
  </div>
</div>
<div class="el-form-item is-required"><label for="injectedProcess" class="el-form-item__label">injection_mapping.injected_process</label>
  <div class="el-form-item__content"><span class="el-cascader"><div class="el-input el-input--suffix"><!----><input type="text" readonly="readonly" autocomplete="off" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon el-icon-arrow-down"></i><!----></span>
    <!----></span>
    <!---->
  </div><span class="el-cascader__label">
      Injected Process 1
    </span></span>
  <!---->
</div>
</div>
<div class="el-form-item">
  <!---->
  <div class="el-form-item__content">
    <div class="process-preview-stub process-preview"></div>
    <!---->
  </div>
</div>
</form> <button type="button" class="el-button el-button--success">
  <!---->
  <!----><span>injection_mapping.save</span></button></div>
`;
